2021_RTCSA_dynfed/src/dynfed.rs

```diff
@@ -48,6 +122,108 @@ where
 
     (minimum_cores, execution_order)
 }
+
+#[allow(dead_code)]
+pub fn dynamic_federated<T>(
+    dag_set: &mut [Graph<NodeData, i32>],
+    processor: &mut T,
+    scheduler: &mut impl SchedulerBase<T>,
+) -> i32
+where
+    T: ProcessorBase + Clone,
+{
+    let mut current_time = 0;
+    let dag_set_length = dag_set.len();
+    let mut dyn_fed_handlers = vec![DynamicFederatedHandler::new(); dag_set_length];
+    let mut dag_queue: VecDeque<Graph<NodeData, i32>> = VecDeque::new();
+    let mut finished_dags_count = 0;
+
+    scheduler.set_processor(processor);
+
+    for (dag_id, dag) in dag_set.iter_mut().enumerate() {
+        dag.set_dag_id(dag_id);
+        dyn_fed_handlers[dag_id].set_minimum_cores_and_execution_order(dag, scheduler);
+    }
+
+    while finished_dags_count < dag_set_length {
+        dag_set
+            .iter_mut()
+            .filter(|dag| current_time == dag.get_head_offset())
+            .for_each(|dag| dag_queue.push_back(dag.clone()));
```

Comment:
It is inefficient for `dat_set` to be traversed every loop and should be improved.

Reply:
Changed to work only at offset times.